doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,252 @@
+from docutils import nodes
+from sphinx.util import logging as sphinx_logging
+
+logger = sphinx_logging.getLogger(__name__)
+
+
+class MissingReferenceFilter(logging.Filter):
+    """
+    A logging filter designed to record missing reference warning messages
+    for use by this extension
+    """
+    def __init__(self, app):
+        self.app = app
+        super(MissingReferenceFilter, self).__init__()
+
+    def _record_reference(self, record):
+        if not getattr(record, 'type', '') == 'ref':
+            return
+
+        if not hasattr(record, 'location'):
+            return
+
+        location = record.location
+        if not isinstance(location, nodes.Node):
+            return
+
+        if not hasattr(self.app.env, "missing_reference_warnings"):
+            self.app.env.missing_reference_warnings = defaultdict(set)
+
+        record_missing_reference(self.app,
+            self.app.env.missing_reference_warnings,
+            record.location)
+
+    def filter(self, record):
+        self._record_reference(record)
+        return True
+
+
+def record_missing_reference(app, record, node):
+    domain = node["refdomain"]
+    typ = node["reftype"]
+    target = node["reftarget"]
+    location = get_location(node, app)
+
+    dtype = "{}:{}".format(domain, typ)
+
+    record[(dtype, target)].add(location)
+
+
+def record_missing_reference_handler(app, env, node, contnode):
+    """
+    When the sphinx app notices a missing reference, it emits an
+    event which calls this function. This function records the missing
+    references for analysis at the end of the sphinx build.
+    """
+    if not app.config.missing_references_enabled:
+        # no-op when we are disabled.
+        return
+
+    if not hasattr(env, "missing_reference_events"):
+        env.missing_reference_events = defaultdict(set)
+
+    record_missing_reference(app, env.missing_reference_events, node)
+
+
+def get_location(node, app):
+    """
+    Given a docutils node and a sphinx application, return a string
+    representation of the source location of this node.
+
+    Usually, this will be of the form "path/to/file:linenumber". Two
+    special values can be emitted, "<external>" for paths which are
+    not contained in this source tree (e.g. docstrings included from
+    other modules) or "<unknown>", inidcating that the sphinx application
+    cannot locate the original source file (usually because an extension
+    has injected text into the sphinx parsing engine).
+    """
+    (path, line) = get_source_line(node)
+
+    if path:
+
+        basepath = os.path.abspath(os.path.join(app.confdir, ".."))
+        path = os.path.relpath(path, start=basepath)
+
+        if path.startswith(os.path.pardir):
+            path = os.path.join("<external>", os.path.basename(path))
+
+    else:
+        path = "<unknown>"
+
+    if line:
+        line = str(line)
+    else:
+        line = ""
+
+    return "%s:%s" % (path, line)
```

Comment:
either pick f-strings or str.format or %-format but stick to one of them throughout :) (well, I'm partial to f-strings)

Reply:
Everywhere uses f-strings now but this one.